libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,575 @@
+  hs.e = genKeyPair()
+  msg &= hs.e.publicKey
+  hs.ss.mixHash(hs.e.publicKey)
+
+template write_s: untyped =
+  trace "noise write s"
+  # Appends EncryptAndHash(s.public_key) to the buffer.
+  msg &= hs.ss.encryptAndHash(hs.s.publicKey)
+
+template dh_ee: untyped =
+  trace "noise dh ee"
+  # Calls MixKey(DH(e, re)).
+  hs.ss.mixKey(dh(hs.e.privateKey, hs.re))
+
+template dh_es: untyped =
+  trace "noise dh es"
+  # Calls MixKey(DH(e, rs)) if initiator, MixKey(DH(s, re)) if responder.
+  when initiator:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+  else:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+
+template dh_se: untyped =
+  trace "noise dh se"
+  # Calls MixKey(DH(s, re)) if initiator, MixKey(DH(e, rs)) if responder.
+  when initiator:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+  else:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+
+template dh_ss: untyped =
+  trace "noise dh ss"
+  # Calls MixKey(DH(s, rs)).
+  hs.ss.mixKey(dh(hs.s.privateKey, hs.rs))
+
+template read_e: untyped =
+  trace "noise read e", size = msg.len
+
+  if msg.len < Curve25519Key.len:
+    raise newException(NoiseHandshakeError, "Noise E, expected more data")
+
+  # Sets re (which must be empty) to the next DHLEN bytes from the message. Calls MixHash(re.public_key).
+  copyMem(addr hs.re[0], addr msg[0], Curve25519Key.len)
+  msg = msg[Curve25519Key.len..msg.high]
+  hs.ss.mixHash(hs.re)
+
+template read_s: untyped =
+  trace "noise read s", size = msg.len
+  # Sets temp to the next DHLEN + 16 bytes of the message if HasKey() == True, or to the next DHLEN bytes otherwise.
+  # Sets rs (which must be empty) to DecryptAndHash(temp).
+  var temp: seq[byte]
+  if hs.ss.cs.hasKey:
+    if msg.len < Curve25519Key.len + ChaChaPolyTag.len:
+      raise newException(NoiseHandshakeError, "Noise S, expected more data")
+   
+    temp.setLen(Curve25519Key.len + 16)
+    copyMem(addr temp[0], addr msg[0], Curve25519Key.len + 16)
+    msg = msg[Curve25519Key.len + 16..msg.high]
+  else:
+    if msg.len < Curve25519Key.len:
+      raise newException(NoiseHandshakeError, "Noise S, expected more data")
+
+    temp.setLen(Curve25519Key.len)
+    copyMem(addr temp[0], addr msg[0], Curve25519Key.len)
+    msg = msg[Curve25519Key.len..msg.high]
+  var plain = hs.ss.decryptAndHash(temp)
+  copyMem(addr hs.rs[0], addr plain[0], Curve25519Key.len)
+
+proc receiveHSMessage(sconn: Connection): Future[seq[byte]] {.async.} =
+  var
+    besize: array[2, byte]
+    size: uint16
+  await sconn.readExactly(addr besize[0], 2)
+  bigEndian16(addr size, addr besize[0])
+  trace "receiveHSMessage", size
+  return await sconn.read(size.int)
+
+proc sendHSMessage(sconn: Connection; buf: seq[byte]) {.async.} =
+  var
+    lesize = buf.len.uint16
+    size: array[2, byte]
+  trace "sendHSMessage", size = lesize
+  bigEndian16(addr size[0], addr lesize)
+  await sconn.write(addr size[0], 2)
+  await sconn.write(buf)
+
+proc packNoisePayload(payload: openarray[byte]): seq[byte] {.inline.} =
```

Comment:
inline is kind of overrated - all it does is place a copy in every module and increase compile times, but a seq-heavy proc like this will not actually get inlined. generally, there's very little reason to use it at all - we can enable lto instead

Reply:
specially for a private proc, it's a noop

Reply:
I'm aware that the simple `inline` prefix is obsolete in C but the fact that it gets copied in every module is exactly what I wanted, that helps LTO and the compiler itself.
Like copy elision, using it out of experience as it often made a difference.
That said, this proc indeed ended being not such a good candidate.